File: src/DesignPattern/Structural/Composite/Composite-go/composite.go
package main

import "fmt"

// Component ...
type Component interface {
	PrintInfo()
	Increase(n int)
	Add(comp Component)
}

// Composite ...
type Composite struct {
	data       int
	components []Component
}

// Add ...
func (p *Composite) Add(comp Component) {
	p.components = append(p.components, comp)
}

// PrintInfo ...
func (p *Composite) PrintInfo() {
	fmt.Printf("Componsite %d\n", p.data)
	for _, item := range p.components {
		item.PrintInfo()
	}
}

// Increase ...
func (p *Composite) Increase(n int) {
	p.data += n
	for _, item := range p.components {
		item.Increase(n)
	}
}

// Leaf ...
type Leaf struct {
	data int
}

// Add ...
func (p *Leaf) Add(comp Component) {
	panic("Leaf not add\n")
}

// PrintInfo ...
func (p *Leaf) PrintInfo() {
	fmt.Printf("Leaf %d\n", p.data)
}

// Increase ...
func (p *Leaf) Increase(n int) {
	p.data += n
}

File: src/DesignPattern/Structural/Composite/Composite-ts/composite.ts
export abstract class Component {
    protected data: number;
    abstract Print();
    abstract Increase();
    abstract Add(compl:Component);
}

export class Composite extends Component {
    private Components: Array<Component>;
    constructor(n: number) {
        super()
        this.data = n;
        this.Components = new Array<Component>();
    }
    public Add(compl:Component){
        this.Components.push(compl);
    }
    public Print() {
        console.log("Composite %d",this.data);
        for (const iterator of this.Components) {
            iterator.Print();
        }
    }
    public Increase() {
        this.data++;
        for (const iterator of this.Components) {
            iterator.Increase();
        }
    }
}

export class Leaf extends Component{
    constructor(n:number){
        super()
        this.data = n;
    }
    public Add(compl:Component){
        throw "叶子类无法添加子类";
    }
    public Increase(){
        this.data++;
    }
    public Print(){
        console.log("Leaf %d",this.data);
    }
}